### ISSUE TITLE
Important: Local Storage Security Breach

### ISSUE THREAD

**Issue by rwieruch** [the-road-to-react-with-firebase/the-road-to-react-with-firebase#15]
Mentioned over here [0][1][2].

- [0] https://www.robinwieruch.de/complete-firebase-authentication-react-tutorial/#comment-4391597595
- [1] https://github.com/the-road-to-react-with-firebase/react-firebase-authentication/issues/21 
- [2] https://github.com/the-road-to-react-with-firebase/react-firebase-authentication/issues/21

Quick Fixes: See my comment in [0] or here is more detail:

- A. remove local storage to prevent this issue entirely, because, as far as I know, it's not possible to temper with the authenticated user coming from Firebase
- B. keep local storage, but act in the withAuthorization HOC only on the real user coming from Firebase and therefore let the blank screen happen if the user is not yet available
- ~C. keep local storage, but implement and trust your security rules, so that even if a user accesses your protected routes, no sensible data is send to the user https://github.com/the-road-to-react-with-firebase/react-firebase-authentication/pull/26~

C wouldn't work with our roles property implementation, because if this is changed maliciously, the security rules implemented as in the PR would fail here as well.

**Comment by here-nerd** [comment]
My opinions 
1. LocalStorage should never be used to store any sensitive information (see https://dev.to/rdegges/please-stop-using-local-storage-1i04) for more details.
2. A frontend logic should never allow access to protected pages by relying only on data stored on the clientside, LocalStorage.
3. Hence i think it is essential to reauthenticate (using authentication state aka access token?) against the server side. So for the flickering issue, we might have just to be creative on the client side eg. show loading..etc.

**Comment by rwieruch** [comment]
Thanks for your input on this @here-nerd 👍 So I think for all the sensible pages we cannot rely on local storage then, but show a loading indicator instead, and only rely on the user coming directly from Firebase. That would mean the withAuthorization HOC needs to be adjusted, because it should only rely on the actual user from Firebase. Otherwise it shows a loading indicator or nothing.

**Comment by here-nerd** [comment]
Yes, that's exactly what I am proposing. Hope it helps.

**Comment by rwieruch** [comment]
Thanks. Am I right to think that people cannot temper with the user coming directly from the Firebase listener? Otherwise people could add roles to this user as well.

**Comment by evdama** [comment]
- The user object you get from firebase-auth contains an ID token when it arrives on the client. You cannot fiddle with this token without firebase-auth taking notice. The token takes the form of a so called JWT.
- This token is send in any request from client to any firebase service e.g. firestore, storeage, hosting... This is how a client request is authenticated inside a backend service.
- Authorization on top of this authentication can be done with so called custom claims which are an object on the ID token. Inside this object you can set roles (and all other kinds of things; KISS though e.g. no more than 1kB etc.)
- When the clients wants to access a firebase resource such as a certain document inside firestore that's only accessible for users with roles admin then you can write a security rule that checks for this claim and either allows the request or denies it.
https://firebase.google.com/docs/auth/admin/custom-claims

Doing Auth & Auth with custom claims is the recommended way, although you could also store roles in a firestore collection that you keep up to date using backend code (e.g. cloud functions). However, compared to the claims approach you'd need several steps until you can either deny or allow a client request for a certain resource whereas with claims you have one request that contains the claim inside the secure ID token. That's why using custom claims inside the ID token is the recommended way.

In fact, I'd refrain from doing Auth & Auth any other way, let alone involve local storage because there you'd have to make sure your solution works on all kinds of user devices, is up to date and secure and if that dealt with you still have a ton of unnecessary code to maintain and drag along. I also believe that rolling your own Auth & Auth has shown to cripple more often than not... there's always that edge case you haven't thought about and that's whey you've got a PR disaster at your hands to deal with. (simpler code, much faster, less requests and last but not least, a lot cheaper as requests cost you money)

Using what Google provides (incl all the security research, maintenance and harmonization as well as field testing), even if it looks like a bit of overkill in the beginning, certainly pays of in the long run...

**Comment by rwieruch** [comment]
Thank you @evdama for writing this up. If I am going to rewrite the book this year with Hooks and Firestore, I will make sure to use the custom claims! Thanks for your thorough explanation on how it works.

Anyway, do you think it's okay regarding security concerns for now by [deploying security rules](https://github.com/the-road-to-react-with-firebase/react-firebase-authentication/pull/26) and [removing the local storage](https://www.robinwieruch.de/react-firebase-auth-persistence/)? It shouldn't be possible to fiddle with the Firebase user, so the roles on the database user should be alright for the security rules, shouldn't they?

**Comment by rwieruch** [comment]
> * Authorization on top of this authentication can be done with so called custom claims which are an object on the ID token. Inside this object you can set roles (and all other kinds of things; KISS though e.g. no more than 1kB etc.)

If I understand the Firebase documentation correctly, these claims are set either via Firebase Cloud Functions or my own Node.js backend with the ADMIN SDK, correct? I am not doing it on the client side.

**Comment by evdama** [comment]
>  It shouldn't be possible to fiddle with the Firebase user, so the roles on the database user should be alright for the security rules, shouldn't they?

If you set the roles using backend code i.e. the admin sdk then yes. No otherwise. Why? Because the admin SDK, while setting claims, unwraps, changes and wraps your user object with the secure ID token, send the new one down to the client(s).

**Comment by rwieruch** [comment]
> > It shouldn't be possible to fiddle with the Firebase user, so the roles on the database user should be alright for the security rules, shouldn't they?
> 
> If you set the roles using backend code i.e. the admin sdk then yes. No otherwise.

Gotcha. Because also the source code to set the roles, e.g. where I merge the Firebase user with the Database user, could be changed, right? So anyone could add code to add the ADMIN role over there.

**Comment by evdama** [comment]
> > It shouldn't be possible to fiddle with the Firebase user, so the roles on the database user should be alright for the security rules, shouldn't they?
> 
> If you set the roles using backend code i.e. the admin sdk then yes. No otherwise.
> 
> Gotcha. Because also the source code to set the roles, e.g. where I merge the Firebase user with the Database user, could be changed, right?

Exactly, you need to do it in a secure environment i.e. one that only you control, one that only runs your trusted code. The admin SDK runs authorized via a service account i.e. has an entirely different Auth & Auth structure than any of the client SDKs.

That kind of Auth & Auth pillar is fundamental to get right. If you do, the sky is the limit really... you're fucked otherwise :)

**Comment by rwieruch** [comment]
Okay, I think I got it. I always thought I get around it, because it seemed like a hassle to set up and the Firebase docs are often without much context. Also wanted to get around the vendor lock-in. 

Now I need to get out a warning to my readers :( Thanks for all your help! I will make sure that the new book has

- [ ] custom claims set via backend code (I guess cloud functions are the easiest)
- [ ] no local storage usage for sensitive data

Optional:

- [ ] React Hooks
- [ ] Cloud Firestore instead of Realtime Database

I always learned tech myself while writing about it. But when it comes to such security concerns, I better get it right for everyone relying on it.

EDIT: Not sure whether you are using the code yourself and adapt it to your security needs, but if you get something working or have just a code snippet for us, let us know about it here :)

**Comment by evdama** [comment]
Love to read this. Security, although everybody thinks isn't hard, is in fact the hardest thing to get right and also the thing that kills business and private finances the fastest if done wrong. That's why in all honesty one should rely on things provided by big well-versed tech players like Google.

Sure, vendor lock-in exists, but then you have it as soon as you get out of bed in the morning... some vendor lock-in is a blessing I guess, the alternative is being a monkey in the jungle ;)

> I always learned tech myself while writing about it. But when it comes to such security concerns, I better get it right for everyone relying on it.

You must. Atm, although the most comprehensive tutorial on fb and react out there, I couldn't recommend it. Change/add the things discussed in this thread and I'd highly recommend it to anyone.

> but if you get something working or have just a code snippet for us, let us know about it here :)

you got the first PR yesterday, others will follow, starting in a month or so... :)

**Comment by evdama** [comment]
roadmap for your book: basics first i.e. you get Auth & Auth done/fixed first, data storage next, then new react APIs such as hooks etc.

... but then I guess you'd do hooks while you go along anyhow, why rewrite two times ;)

**Comment by tpperlman** [comment]
Does this issue affect Ionic 4 apps built with Capacitor and compiled for Android/iOS?

**Comment by DevelopmentIntensified** [comment]
> * [ ]  custom claims set via backend code (I guess cloud functions are the easiest)
> * [ ]  no local storage usage for sensitive data

instead of using custom claims for admin only updatable values, you could use security rules on the database to make it so that only a user that is already an admin can change the value of the admin value. This method is much easier than setting up cloud functions which require a pay as you go plan. I am not sure if this works with the real-time database however with firebase I was able to do it using: https://firebase.google.com/docs/firestore/security/rules-fields?authuser=0#forbidding_specific_fields_in_new_documents.
